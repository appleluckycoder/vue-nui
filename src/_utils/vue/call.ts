export function call<A extends any[]> (funcs: Function[] | Function, ...args: A): void {
  if (Array.isArray(funcs)) funcs.forEach((func) => call(func, ...args))
  else return funcs(...args)
}

