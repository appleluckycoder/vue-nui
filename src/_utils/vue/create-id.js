export function createId (length = 8) {
  return Math.random().toString(16).slice(2, 2 + length)
}
