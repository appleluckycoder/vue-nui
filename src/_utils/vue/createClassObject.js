export default function createClassObject (classValue) {
  if (typeof classValue === 'string') {
    return classValue
      .split(' ')
      .filter(v => v.length)
      .reduce((prevValue, currentValue) => {
        prevValue[currentValue] = true
        return prevValue
      }, {})
  } else if (classValue && typeof classValue === 'object') {
    return classValue
  }
  return null
}
