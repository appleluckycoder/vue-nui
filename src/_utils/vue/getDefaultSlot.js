export default function (instance, fallback) {
  fallback = fallback === undefined ? [] : fallback
  const defaultScopedSlot = instance.$slots.default
  return (defaultScopedSlot && defaultScopedSlot()) || fallback
}
