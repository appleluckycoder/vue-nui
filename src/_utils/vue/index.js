export * from './src/get-slot'
export * from './src/get-v-node-children'
export * from './src/create-id'
