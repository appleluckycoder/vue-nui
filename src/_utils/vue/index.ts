export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export type { MaybeArray } from './call'
