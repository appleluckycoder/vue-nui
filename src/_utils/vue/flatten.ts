import { Fragment, VNodeChild } from 'vue'

// o(n) flatten
export function flatten (vNodes: VNodeChild[], result: VNodeChild[] = []) {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      result.push(vNode)
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, result)
      }
      // rawSlot
    } else {
      result.push(vNode)
    }
  })
  return result
}
