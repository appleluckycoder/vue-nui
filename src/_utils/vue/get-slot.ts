import { ComponentPublicInstance, VNode } from "vue"

export function getSlot (instance: ComponentPublicInstance, slotName = 'default', fallback: VNode[] = []) {
  const slots = instance.$slots
  const slot = slots[slotName]
  if (slot === undefined) return fallback
  return slot()
}
