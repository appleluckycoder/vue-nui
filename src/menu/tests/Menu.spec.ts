import { mount } from '@vue/test-utils'
import { h } from 'vue'
import { sleep } from 'seemly'
import { NMenu } from '../index'

describe('n-menu', () => {
  it('should work with import on demand', () => {
    mount(NMenu)
  })
  it('props.onUpdateValue type', () => {
    const stringCb = (v: string): void => {}
    const numberCb = (v: number): void => {}
    const snCb = (v: string | number): void => {}
    const stringArrCb = (v: string[]): void => {}
    const numberArrCb = (v: number[]): void => {}
    const snArrCb = (v: Array<string | number>): void => {}
    mount(NMenu, {
      props: {
        onUpdateValue: stringCb,
        onUpdateExpandedKeys: stringArrCb
      }
    })
    mount(NMenu, {
      props: {
        onUpdateValue: numberCb,
        onUpdateExpandedKeys: numberArrCb
      }
    })
    mount(NMenu, {
      props: {
        onUpdateValue: snCb,
        onUpdateExpandedKeys: snArrCb
      }
    })
  })
  it('should tooltip work with `render-label` props', async () => {
    const options = [
      {
        label: () =>
          h(
            'a',
            {
              href: 'test1',
              target: '_blank',
              rel: 'test1'
            },
            'test1'
          ),
        key: 'test1'
      },
      {
        label: 'test2',
        key: 'test2'
      }
    ]
    const renderLabel = (option: any): any => {
      if (typeof option.label === 'function') {
        return option.label()
      }
      return h(
        'a',
        {
          href: option.key,
          rel: option.key
        },
        { default: () => option.label }
      )
    }
    const wrapper = mount(NMenu, {
      props: {
        options: options
      }
    })
    expect(wrapper.find('[href="test1"]').exists()).toBe(true)
    expect(wrapper.find('[href="test2"]').exists()).toBe(false)

    await wrapper.setProps({ renderLabel: renderLabel })
    expect(wrapper.find('[href="test1"]').exists()).toBe(true)
    expect(wrapper.find('[target="_blank"]').exists()).toBe(true)
    expect(wrapper.find('[href="test2"]').exists()).toBe(true)
  })
  it('should dropdown work with `render-label` props', async () => {
    const options = [
      {
        label: 'jj',
        key: 'jj'
      },
      {
        label: 'jay',
        key: 'jay',
        children: [
          {
            type: 'group',
            label: 'song-group',
            key: 'group',
            children: [
              {
                label: 'fantasy',
                key: 'fantasy'
              },
              {
                label: 'mojito',
                key: 'mojito'
              }
            ]
          }
        ]
      }
    ]
    const renderLabel = (option: any): any => {
      return h(
        'a',
        {
          class: option.key,
          href: option.key,
          rel: option.key
        },
        { default: () => option.label }
      )
    }
    const wrapper = mount(NMenu, {
      props: {
        options: options,
        collapsed: true,
        renderLabel: renderLabel
      }
    })
    expect(wrapper.find('.n-submenu').exists()).toBe(true)
    await wrapper.find('.n-submenu').trigger('mouseenter')
    // Popover has delay, so we need to wait
    await sleep(150)
    expect(document.body.querySelector('.n-dropdown')).not.toEqual(null)
    expect(document.querySelectorAll('a').length).toEqual(3)
    expect(document.querySelectorAll('a.fantasy').length).toEqual(1)
  })
})
