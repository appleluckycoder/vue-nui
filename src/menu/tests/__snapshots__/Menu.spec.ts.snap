// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-menu should dropdown work with \`render-label\` props 1`] = `
<body>
  <div
    class="v-binder-follower-container"
    style="z-index: 2000;"
  >
    <div
      class="v-binder-follower-content"
    >
      <transition-stub>
        <div
          class="n-dropdown-menu n-popover n-dropdown"
          style="--box-shadow: 0 3px 6px -4px rgba(0, 0, 0, .12), 0 6px 16px 0 rgba(0, 0, 0, .08), 0 9px 28px 8px rgba(0, 0, 0, .05); --bezier: cubic-bezier(.4, 0, .2, 1); --bezier-ease-in: cubic-bezier(.4, 0, 1, 1); --bezier-ease-out: cubic-bezier(0, 0, .2, 1); --font-size: 14px; --text-color: rgb(51, 54, 57); --color: #fff; --divider-color: rgb(239, 239, 245); --border-radius: 3px; --arrow-height: 6px; --arrow-offset: 10px; --arrow-offset-vertical: 10px; --padding: 4px 0; --space: 6px; --space-arrow: 10px; --option-height: 40px; --option-prefix-width: 16px; --option-icon-prefix-width: 40px; --option-suffix-width: 16px; --option-icon-suffix-width: 36px; --option-icon-size: 18px; --option-opacity-disabled: 0.5; --option-color-hover: rgb(243, 243, 245); --option-color-active: rgba(24, 160, 88, 0.1); --option-text-color: rgb(51, 54, 57); --option-text-color-hover: rgb(51, 54, 57); --option-text-color-active: #18a058; --option-text-color-child-active: #18a058; --prefix-color: rgb(51, 54, 57); --suffix-color: rgb(51, 54, 57); --group-header-text-color: rgb(158, 164, 170);"
        >
          
          <div
            class="n-dropdown-option"
          >
            <div
              class="n-dropdown-option-body n-dropdown-option-body--group"
            >
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__prefix"
              />
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__label"
              >
                song-group
              </div>
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__suffix"
              />
            </div>
          </div>
          
          <div
            class="n-dropdown-option"
          >
            <div
              class="n-dropdown-option-body"
            >
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__prefix"
              />
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__label"
              >
                <a
                  class="fantasy"
                  href="fantasy"
                  rel="fantasy"
                >
                  fantasy
                </a>
              </div>
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__suffix"
              />
            </div>
            <!---->
          </div>
          <div
            class="n-dropdown-option"
          >
            <div
              class="n-dropdown-option-body"
            >
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__prefix"
              />
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__label"
              >
                <a
                  class="mojito"
                  href="mojito"
                  rel="mojito"
                >
                  mojito
                </a>
              </div>
              <div
                __dropdown-option="true"
                class="n-dropdown-option-body__suffix"
              />
            </div>
            <!---->
          </div>
          
          
        </div>
      </transition-stub>
    </div>
  </div>
  
</body>
`;
